cpp/src/arrow/array/builder_dict.h

```diff
@@ -282,6 +294,163 @@ class DictionaryBuilderBase : public ArrayBuilder {
+        }
+        break;
+      }
+      case Type::INT8: {
+        const auto& value = dict.GetView(
+            internal::checked_cast<const Int8Scalar&>(*dict_scalar.value.index).value);
+        for (int64_t i = 0; i < n_repeats; i++) {
+          ARROW_RETURN_NOT_OK(Append(value));
+        }
+        break;
+      }
+      case Type::UINT16: {
+        const auto& value = dict.GetView(
+            internal::checked_cast<const UInt16Scalar&>(*dict_scalar.value.index).value);
+        for (int64_t i = 0; i < n_repeats; i++) {
+          ARROW_RETURN_NOT_OK(Append(value));
+        }
+        break;
+      }
+      case Type::INT16: {
+        const auto& value = dict.GetView(
+            internal::checked_cast<const Int16Scalar&>(*dict_scalar.value.index).value);
+        for (int64_t i = 0; i < n_repeats; i++) {
+          ARROW_RETURN_NOT_OK(Append(value));
+        }
+        break;
+      }
+      case Type::UINT32: {
+        const auto& value = dict.GetView(
+            internal::checked_cast<const UInt32Scalar&>(*dict_scalar.value.index).value);
+        for (int64_t i = 0; i < n_repeats; i++) {
+          ARROW_RETURN_NOT_OK(Append(value));
+        }
+        break;
+      }
+      case Type::INT32: {
+        const auto& value = dict.GetView(
+            internal::checked_cast<const Int32Scalar&>(*dict_scalar.value.index).value);
+        for (int64_t i = 0; i < n_repeats; i++) {
+          ARROW_RETURN_NOT_OK(Append(value));
+        }
+        break;
+      }
+      case Type::UINT64: {
+        const auto& value = dict.GetView(
+            internal::checked_cast<const UInt64Scalar&>(*dict_scalar.value.index).value);
+        for (int64_t i = 0; i < n_repeats; i++) {
+          ARROW_RETURN_NOT_OK(Append(value));
+        }
+        break;
+      }
+      case Type::INT64: {
+        const auto& value = dict.GetView(
+            internal::checked_cast<const Int64Scalar&>(*dict_scalar.value.index).value);
+        for (int64_t i = 0; i < n_repeats; i++) {
+          ARROW_RETURN_NOT_OK(Append(value));
+        }
+        break;
+      }
+      default:
+        return Status::TypeError("Invalid index type: ", dict_ty);
+    }
+    return Status::OK();
+  }
+
+  Status AppendScalars(const ScalarVector& scalars) override {
+    for (const auto& scalar : scalars) {
+      ARROW_RETURN_NOT_OK(AppendScalar(*scalar, /*n_repeats=*/1));
+    }
+    return Status::OK();
+  }
+
+  Status AppendArraySlice(const ArrayData& array, int64_t offset, int64_t length) final {
+    // Visit the indices and insert the unpacked values.
+    const auto& dict_ty = internal::checked_cast<const DictionaryType&>(*array.type);
+    const typename TypeTraits<T>::ArrayType dict(array.dictionary);
+    switch (dict_ty.index_type()->id()) {
+      case Type::UINT8: {
+        const uint8_t* values = array.GetValues<uint8_t>(1) + offset;
+        return VisitBitBlocks(
+            array.buffers[0], array.offset + offset, std::min(array.length, length),
```

Comment:
Other `AppendArraySlice` implementations don't check that `length` is in bounds, so `std::min` doesn't seem necessary here.